core/file.go

```diff
@@ -509,22 +509,54 @@ func (file *File) WithTimestamps(ctx context.Context, unix int) (*File, error) {
 	}, withSavedSnapshot("withTimestamps %d", unix))
 }
 
+type fileReadCloser struct {
+	read  func(p []byte) (n int, err error)
+	close func() error
+}
+
+func (frc fileReadCloser) Read(p []byte) (n int, err error) {
+	return frc.read(p)
+}
+
+func (frc fileReadCloser) Close() error {
+	return frc.close()
+}
+
+var _ io.ReadCloser = fileReadCloser{}
+
 func (file *File) Open(ctx context.Context) (io.ReadCloser, error) {
-	query, err := CurrentQuery(ctx)
+	root, closer, err := execInMountCloser(ctx, file, allowNilBuildkitSession)
 	if err != nil {
 		return nil, err
 	}
-	bk, err := query.Buildkit(ctx)
+
+	filePath, err := containerdfs.RootPath(root, file.File)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get buildkit client: %w", err)
+		_, _ = closer(true)
+		return nil, err
 	}
 
-	fs, err := reffs.OpenDef(ctx, bk, file.LLB)
```

Comment:
Oof, this is such a cursed function lol.

Very tempted to suggest to keep pulling on this thread, and rip it out entirely - it also seems to be used in `resolveUIDGID` which seems like it could use the mount ref helpers.